src/main/java/org/cryptomator/windows/common/RegistryKey.java

```diff
@@ -0,0 +1,171 @@
+
+import static java.lang.foreign.MemorySegment.NULL;
+import static org.cryptomator.windows.capi.common.Windows_h.ERROR_FILE_NOT_FOUND;
+import static org.cryptomator.windows.capi.common.Windows_h.ERROR_MORE_DATA;
+import static org.cryptomator.windows.capi.common.Windows_h.ERROR_SUCCESS;
+import static org.cryptomator.windows.capi.winreg.Winreg_h.REG_EXPAND_SZ;
+import static org.cryptomator.windows.capi.winreg.Winreg_h.RRF_RT_REG_DWORD;
+import static org.cryptomator.windows.capi.winreg.Winreg_h.RRF_RT_REG_EXPAND_SZ;
+import static org.cryptomator.windows.capi.winreg.Winreg_h.RRF_RT_REG_SZ;
+
+public class RegistryKey implements AutoCloseable {
+
+	public static final RegistryKey HKEY_CURRENT_USER = new RegistryKey(Winreg_h.HKEY_CURRENT_USER(), "HKEY_CURRENT_USER");
+	public static final RegistryKey HKEY_LOCAL_MACHINE = new RegistryKey(Winreg_h.HKEY_LOCAL_MACHINE(), "HKEY_LOCAL_MACHINE");
+	public static final RegistryKey HKEY_CLASSES_ROOT = new RegistryKey(Winreg_h.HKEY_CLASSES_ROOT(), "HKEY_CLASSES_ROOT");
+	public static final RegistryKey HKEY_USERS = new RegistryKey(Winreg_h.HKEY_USERS(), "HKEY_USERS");
+
+	private final String path;
+	private MemorySegment handle;
+	private volatile boolean isClosed = false;
+
+	RegistryKey(MemorySegment handle, String path) {
+		this.handle = handle;
+		this.path = path;
+	}
+
+	//-- GetValue functions --
+
+	public String getStringValue(String name, boolean isExpandable) throws RegistryValueException {
+		try (var arena = Arena.ofConfined()) {
+			var lpValueName = arena.allocateFrom(name, StandardCharsets.UTF_16LE);
+			var data = getValue(lpValueName, isExpandable ? RRF_RT_REG_EXPAND_SZ() : RRF_RT_REG_SZ(), 256L);
+			return data.getString(0, StandardCharsets.UTF_16LE);
+		}
+	}
+
+	public int getDwordValue(String name) throws RegistryValueException {
+		try (var arena = Arena.ofConfined()) {
+			var lpValueName = arena.allocateFrom(name, StandardCharsets.UTF_16LE);
+			var data = getValue(lpValueName, RRF_RT_REG_DWORD(), 5L);
+			return data.get(ValueLayout.JAVA_INT, 0);
+		}
+	}
+
+	private MemorySegment getValue(MemorySegment lpValueName, int dwFlags, long seed) throws RegistryValueException {
+		long bufferSize = seed - 1;
+		try (var arena = Arena.ofConfined()) {
+			var lpDataSize = arena.allocateFrom(ValueLayout.JAVA_INT, (int) bufferSize);
+
+			try (var dataArena = Arena.ofConfined()) {
+				var lpData = dataArena.allocate(bufferSize);
+
+				int result = Winreg_h.RegGetValueW(handle, NULL, lpValueName, dwFlags, NULL, lpData, lpDataSize);
+				if (result == ERROR_MORE_DATA()) {
+					throw new BufferTooSmallException();
+				} else if (result != ERROR_SUCCESS()) {
+					throw new RegistryValueException("winreg_h:RegGetValue", path, lpValueName.getString(0, StandardCharsets.UTF_16LE), result);
+				}
+
+				var returnBuffer = Arena.ofAuto().allocate(Integer.toUnsignedLong(lpDataSize.get(ValueLayout.JAVA_INT, 0)));
+				MemorySegment.copy(lpData, 0L, returnBuffer, 0L, returnBuffer.byteSize());
```

Comment:
no need for multiple confined arenas. Why not pass in the arena from the caller? This would also eliminate the necessity to copy the result data between segments with different lifecycles.

Reply:
Memory segments can only be released/freed, if the arena is closed. Since we exponentially increase the buffer size in every try, we allocate a lot of memory anyway. And if we do not use a separate arena for each try, for every ERROR_MORE_DATA we allocate more memory without freeing the old one, leading to a huge amount of allocated memory.

I rewrote the method to use a while loop and added some comments in https://github.com/cryptomator/integrations-win/pull/69/commits/b2856be9ce5d4b2a3ba108ca084eab15ca4ab36c

Reply:
Let's have a short call, as this is not what I had in mind.

Reply:
In a session with @overheadhunter we simplified the method (and lowered the maximum data size to set/get for reg values). Commited in 35fcd7716381880c1c4caeecd5fd2bae4388a2d0